File: Find_nearBiggest.m
function [HR,whichPeak] = Find_nearBiggest(Fre, HR_prev, rangeplus,rangeminus)
    %找离HR_prev一定范围内最大的峰位置,前6个里面找最大的
    %Fre：峰位置集合，根据幅值降序排列
    %HR_prev：上一时刻心率
    %range：允许的心率变化范围
       
    HR = HR_prev;
    whichPeak = 0;
    
    if length(Fre) > 5
        len = 5 ;
    else
        len = length(Fre);
    end
    
        for i = 1:len
           if (Fre(i)- HR_prev < rangeplus)  && ( Fre(i) - HR_prev > rangeminus)
               HR = Fre(i);
               whichPeak = i;
               break
           end
        end
%         HR = Fre(1);


end

File: python/src/ppg_hr/core/find_near_biggest.py
"""Locate the first peak in a small window near the previous HR (port of ``Find_nearBiggest.m``).

Among the first 5 amplitude-sorted candidates, return the first that falls
strictly inside ``(hr_prev + range_minus, hr_prev + range_plus)``. If none
qualifies, return ``hr_prev`` and ``which_peak = 0``. ``which_peak`` is
returned in 1-based MATLAB indexing so callers can compare directly against
MATLAB code.
"""

from __future__ import annotations

import numpy as np

__all__ = ["find_near_biggest"]


def find_near_biggest(
    fre: np.ndarray,
    hr_prev: float,
    range_plus: float,
    range_minus: float,
) -> tuple[float, int]:
    """Return ``(hr, which_peak_1based)``."""
    arr = np.atleast_1d(np.asarray(fre, dtype=float)).ravel()
    n = arr.size
    limit = min(5, n)
    for i in range(limit):
        delta = arr[i] - hr_prev
        if (delta < range_plus) and (delta > range_minus):
            return float(arr[i]), i + 1
    return float(hr_prev), 0
